packages/concerto-core/lib/introspect/mapkeytype.js

```diff
@@ -0,0 +1,136 @@
+
+const Decorated = require('./decorated');
+const IllegalModelException = require('./illegalmodelexception');
+
+// Types needed for TypeScript generation.
+/* eslint-disable no-unused-vars */
+/* istanbul ignore next */
+if (global === undefined) {
+    const ModelFile = require('./modelfile');
+}
+
+/**
+ * MapKeyType defines a Key type of an MapDeclaration.
+ *
+ * @extends Decorated
+ * @see See {@link Decorated}
+ * @class
+ * @memberof module:concerto-core
+ */
+class MapKeyType extends Decorated {
+    /**
+     * Create an MapKeyType.
+     * @param {MapDeclaration} parent - The owner of this property
+     * @param {Object} ast - The AST created by the parser
+     * @param {ModelFile} modelFile - the ModelFile for the Map class
+     * @throws {IllegalModelException}
+     */
+    constructor(parent, ast) {
+        super(ast);
+        this.parent = parent;
+        this.type = null;
+        this.process();
+    }
+
+    /**
+     * Process the AST and build the model
+     *
+     * @throws {IllegalModelException}
+     * @private
+     */
+    process() {
+        super.process();
+        this.type = this.ast.name;
+    }
+
+    /**
+     * Semantic validation of the structure of this class.
+     *
+     * @throws {IllegalModelException}
+     * @protected
+     */
+    validate() {
+        const declaration = this.getModelFile().getAllDeclarations();
+        const key = declaration.find(decl => decl.name === this.type);
+
+        if (!key?.isConcept?.()           &&
+            !key?.isEnum?.()              &&
+            !key?.isScalarDeclaration?.() &&
+            !['String', 'Boolean', 'DateTime'].includes(this.type)) {
+            throw new IllegalModelException(`MapKeyType has invalid Type: ${this.type}`);
+        }
+
+        if (key?.isConcept?.() && !key.isIdentified()) {
+            throw new IllegalModelException(
+                `ConceptDeclaration must be identified in context of MapKeyType: ${this.type}`
+            );
+        }
+
+        if (key?.isScalarDeclaration?.() &&
+            !(key?.ast.$class === `${MetaModelNamespace}.StringScalar`)  &&
+            !(key?.ast.$class === `${MetaModelNamespace}.BooleanScalar`) &&
+            !(key?.ast.$class === `${MetaModelNamespace}.DateTimeScalar` )) {
+            throw new IllegalModelException(
+                `Scalar must be one of StringScalar, BooleanScalar, DateTimeScalar in context of MapKeyType. Invalid Scalar: ${this.type}`
+            );
+        }
+
+    }
+
+    /**
+     * Returns the ModelFile that defines this class.
+     *
+     * @public
+     * @return {ModelFile} the owning ModelFile
+     */
+    getModelFile() {
+        return this.parent.getModelFile();
+    }
+
+    /**
+     * Returns the name of the MapKey. This name does not include the
+     * namespace from the owning ModelFile.
+     *
+     * @return {string} the short name of this class
+     */
+    getName() {
+        return this.ast.name;
+    }
```

Comment:
The AST doesn't have a property called `name` AFAICT?

Reply:
You are correct. Its a little ambiguous, the AST in this case is a subset representing the MapDeclaration only. There is no need to process the entire AST, the constructor only uses what it needs from the parse tree.

```
{
  '$class': 'concerto.metamodel@1.0.0.MapDeclaration',
  name: 'Dictionary',
  properties: [
    {
      '$class': 'concerto.metamodel@1.0.0.MapKeyType',
      name: 'Person',
      location: [Object]
    },
    {
      '$class': 'concerto.metamodel@1.0.0.AggregateValueType',
      name: 'String',
      location: [Object]
    }
  ],
  location: {
    '$class': 'concerto.metamodel@1.0.0.Range',
    start: {
      offset: 676,
      line: 19,
      column: 1,
      '$class': 'concerto.metamodel@1.0.0.Position'
    },
    end: {
      offset: 716,
      line: 22,
      column: 2,
      '$class': 'concerto.metamodel@1.0.0.Position'
    }
  }
} 
```